cmd/routed-eni-cni-plugin/driver/driver.go

```diff
@@ -356,6 +471,35 @@ func tearDownNS(addr *net.IPNet, table int, netLink netlinkwrapper.NetLink, log
 	return nil
 }
 
+// TeardownPodENINetwork tears down the vlan and corresponding ip rules.
+func (os *linuxNetwork) TeardownPodENINetwork(vlanId int, log logger.Logger) error {
+	log.Infof("Tear down of pod ENI namespace")
+
+	// 1. delete vlan
+	if vlan, err := os.netLink.LinkByName(fmt.Sprintf("vlan.eth.%d",
+		vlanId)); err == nil {
+		err := os.netLink.LinkDel(vlan)
+		if err != nil {
+			return errors.Wrapf(err, "TeardownPodENINetwork: failed to delete vlan link for %d", vlanId)
+		}
+	}
+
+	// 2. delete two ip rules associated with the vlan
+	vlanRule := os.netLink.NewRule()
+	vlanRule.Table = vlanId + 100
+	vlanRule.Priority = vlanRulePriority
+
+	for {
+		if err := os.netLink.RuleDel(vlanRule); err != nil {
+			if !containsNoSuchRule(err) {
+				return errors.Wrapf(err, "TeardownPodENINetwork: failed to delete container rule for %d", vlanId)
+			}
+			break
+		}
+	}
```

Comment:
This is slightly tricky and it would be nice to have a comment explaining how it works. Basically, we create a new route table for each vlan in` SetupPodENINetwork()`. The table only has two rules, defined in `buildRoutesForVlan()`. When we tear down, we want to delete both those rules. Calling `netLink.RuleDel` with only the table name and priority will delete the first rule matching that, that's why we have the for-loop.